Fix issues with login status

Yicong-Huang opened the pull request:
This PR fixed the following issues 

- Fixed #1008 an un-login user can still access workflow in cache. 
- Fixed #1023 after switching user, workflow sometimes cannot be loaded.

The reason of those issues are simple: user change event is not detected or checked properly.

zuozhiw commented on core/new-gui/src/app/common/util/predicate.ts:
```diff
@@ -0,0 +1,7 @@
+ * @param x
+ * @returns {boolean}
+ */
+export function isDefined(x: any): boolean {return x !== undefined; }
```
any reason we need such a trivial function here? also it's better to add a type guard

Yicong-Huang replied:
This is just used as a predicate lambda. I don't want to write plain `xxx.filter((x)=>{x!==undefined})`. For the type guard, I think it can take in any value, do I need to change?

zuozhiw commented on core/new-gui/src/app/common/service/user/user.service.ts:
```diff
@@ -66,8 +67,8 @@ export class UserService {
-  public getUser(): User|undefined {
-    return this.currentUser;
+  public getUser(): Observable<User|undefined> {
+    return of(this.currentUser);
```
don't we already have another observable below?
```
 public userChanged(): Observable<User|undefined> {
```
If you want the observable to emit undefined in the beginning, you can either add a `startsWith` to the observable, or you can use a BehaviorSubject (this might fit your use case better) https://www.learnrxjs.io/learn-rxjs/subjects/behaviorsubject

Yicong-Huang replied:
changed to use `ReplaySubject` which replays one last value.

zuozhiw commented on core/new-gui/src/app/workspace/component/workspace.component.ts:
```diff
@@ -162,11 +169,17 @@ export class WorkspaceComponent implements OnInit {
+  private noAccessToWorkflow(): void {
+    this.location.go(`/`);
+    // TODO: replace with a proper error message with the framework
+    alert('You don\'t have access to this workflow, please log in with another account');
```
if there's no access, do we need to clear the workflow data? I don't see the code that clears the workflow. I saw we are still reloading the cached workflow. Or does location.go() solves it?

Yicong-Huang replied:
no. `location.go()` only changes url in address bar (location). it does not load any components. The clear workflow code was outside this method, now I moved them inside.

zuozhiw reviewed (changes requested):
left some comments

zuozhiw reviewed (approved):
1. see my comments on `isDefined` function
2. revert the change `package-lock.json` in this PR, usually  `package-lock.json` shouldn't be part of PR if `package.json` is not changed.

Yicong-Huang commented:
> * see my comments on `isDefined` function
> * revert the change `package-lock.json` in this PR, usually  `package-lock.json` shouldn't be part of PR if `package.json` is not changed.

1. thanks. changed to user-defined type guards;
2. not sure why `package-lock.json` is updated here. reverted.
